File: samples/java/Bingo/Server/Session/src/main/java/systems/zlink/samples/bingo/server/session/sessions/BingoSession.java
package systems.zlink.samples.bingo.server.session.sessions;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import systems.zlink.framework.messaging.ZLinkMessage;
import systems.zlink.framework.streams.ZLinkSession;
import systems.zlink.framework.streams.ZLinkSessionActor;
import systems.zlink.framework.streams.ZLinkSessionContext;
import systems.zlink.framework.streams.ZLinkSessionDispatchContext;
import systems.zlink.framework.streams.ZLinkSessionPacketDispatcher;
import systems.zlink.framework.streams.ZLinkStreamError;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public final class BingoSession implements ZLinkSession {
    private static final Logger logger = LoggerFactory.getLogger(BingoSession.class);

    private final ZLinkSessionContext context;
    private final ZLinkSessionPacketDispatcher<ZLinkSessionContext> handlers;
    private String boundActorId;

    public BingoSession(
            ZLinkSessionContext context,
            ZLinkSessionPacketDispatcher<ZLinkSessionContext> handlers) {
        this.context = context;
        this.handlers = handlers;
    }

    @Override
    public ZLinkSessionContext context() {
        return context;
    }

    @Override
    public CompletionStage<Void> onConnected() {
        return CompletableFuture.completedFuture(null);
    }

    // --8<-- [start:doc-bingo-session-disconnect]
    @Override
    public CompletionStage<Void> onDisconnected() {
        // Framework cleanup owns disconnect notification; this callback only records
        // the sample lifecycle evidence without submitting another notification.
        if (boundActorId != null) {
            logger.info("bingo-lifecycle session-disconnect actor={} destroy=false", boundActorId);
        }
        return CompletableFuture.completedFuture(null);
    }

    // --8<-- [end:doc-bingo-session-disconnect]

    @Override
    public CompletionStage<Void> onError(ZLinkStreamError error) {
        return CompletableFuture.completedFuture(null);
    }

    // --8<-- [start:doc-bingo-session-relay]
    @Override
    public CompletionStage<Void> onDispatch(
            ZLinkSessionDispatchContext dispatch, ZLinkMessage payload) {
        return handlers.tryHandle(context, dispatch, payload)
                .thenCompose(
                        handled ->
                                handled
                                        ? rememberBoundActor()
                                        : requireSingleBoundActor(dispatch.packetName())
                                                .relay(payload)
                                                .thenApply(ignored -> null));
    }

    // --8<-- [end:doc-bingo-session-relay]

    private CompletionStage<Void> rememberBoundActor() {
        if (context.actors().bound().size() == 1) {
            boundActorId = context.actors().bound().getFirst().actorId();
        }
        return CompletableFuture.completedFuture(null);
    }

    private ZLinkSessionActor requireSingleBoundActor(String packetName) {
        return switch (context.actors().bound().size()) {
            case 1 -> context.actors().bound().get(0);
            case 0 ->
                    throw new IllegalStateException(
                            "Client must authenticate before relaying packet '" + packetName + "'");
            default ->
                    throw new IllegalStateException(
                            "Exactly one actor must be bound before relaying packet '"
                                    + packetName
                                    + "'");
        };
    }
}

File: samples/kotlin/Bingo/Server/Session/src/main/kotlin/systems/zlink/samples/kotlin/bingo/server/session/sessions/BingoSession.kt
package systems.zlink.samples.kotlin.bingo.server.session.sessions

import kotlinx.coroutines.future.await
import org.slf4j.LoggerFactory
import systems.zlink.framework.kotlin.ZLinkSuspendingSession
import systems.zlink.framework.messaging.ZLinkMessage
import systems.zlink.framework.streams.ZLinkSessionActor
import systems.zlink.framework.streams.ZLinkSessionContext
import systems.zlink.framework.streams.ZLinkSessionDispatchContext
import systems.zlink.framework.streams.ZLinkSessionPacketDispatcher

class BingoSession(
    private val context: ZLinkSessionContext,
    private val handlers: ZLinkSessionPacketDispatcher<ZLinkSessionContext>,
) : ZLinkSuspendingSession() {
    private val logger = LoggerFactory.getLogger(BingoSession::class.java)
    private var boundActorId: String? = null

    override fun context(): ZLinkSessionContext = context

    // --8<-- [start:doc-bingo-session-disconnect]
    override suspend fun onDisconnectedSuspending() {
        // Framework cleanup owns disconnect notification; this callback only records
        // the sample lifecycle evidence without submitting another notification.
        boundActorId?.let { actorId ->
            logger.info("bingo-lifecycle session-disconnect actor={} destroy=false", actorId)
        }
    }

    // --8<-- [end:doc-bingo-session-disconnect]

    // --8<-- [start:doc-bingo-session-relay]
    override suspend fun onDispatchSuspending(
        dispatch: ZLinkSessionDispatchContext,
        payload: ZLinkMessage,
    ) {
        val handled = handlers.tryHandle(context, dispatch, payload).await()
        if (handled) {
            if (context.actors().bound().size == 1) {
                boundActorId = context.actors().bound()[0].actorId()
            }
            return
        }
        val actor = requireSingleBoundActor(dispatch.packetName())
        actor.relay(dispatch, payload).await()
    }

    // --8<-- [end:doc-bingo-session-relay]

    private fun requireSingleBoundActor(packetName: String): ZLinkSessionActor =
        when (context.actors().bound().size) {
            1 -> context.actors().bound()[0]
            0 ->
                throw IllegalStateException(
                    "Client must authenticate before relaying packet '$packetName'"
                )
            else ->
                throw IllegalStateException(
                    "Exactly one actor must be bound before relaying packet '$packetName'"
                )
        }
}
